crates/refund/src/tests.rs

```diff
@@ -55,3 +55,59 @@ fn test_refund_succeeds() {
         assert_ok!(Refund::_execute_refund(refund_id, vec![0u8; 100], vec![0u8; 100],));
     })
 }
+
+mod spec_based_tests {
+    use crate::{RefundRequest, RefundRequests};
+
+    use super::*;
+
+    #[test]
+    fn precondition_refund_must_exist() {
+        run_test(|| {
+            assert_err!(
+                Refund::_execute_refund(H256::random(), vec![0u8; 100], vec![0u8; 100]),
+                TestError::RefundIdNotFound,
+            );
+        })
+    }
+
+    #[test]
+    fn precondition_refund_must_not_be_completed() {
+        run_test(|| {
+            let redeem_id = H256::random();
+            <RefundRequests<Test>>::insert(
+                redeem_id,
+                RefundRequest {
+                    completed: true,
+                    ..Default::default()
+                },
+            );
+            assert_err!(
+                Refund::_execute_refund(redeem_id, vec![0u8; 100], vec![0u8; 100]),
+                TestError::RefundCompleted,
+            );
+        })
+    }
+
+    #[test]
+    fn postcondition_refund_must_be_completed() {
+        run_test(|| {
+            ext::btc_relay::parse_merkle_proof::<Test>.mock_safe(|_| MockResult::Return(Ok(dummy_merkle_proof())));
+            ext::btc_relay::parse_transaction::<Test>.mock_safe(|_| MockResult::Return(Ok(Transaction::default())));
+            ext::btc_relay::verify_and_validate_op_return_transaction::<Test, Balance>
+                .mock_safe(|_, _, _, _, _| MockResult::Return(Ok(())));
+            ext::vault_registry::try_increase_to_be_issued_tokens::<Test>.mock_safe(|_, _| MockResult::Return(Ok(())));
+            ext::vault_registry::issue_tokens::<Test>.mock_safe(|_, _| MockResult::Return(Ok(())));
+
+            let redeem_id = H256::random();
+            <RefundRequests<Test>>::insert(redeem_id, RefundRequest { ..Default::default() });
+
+            assert_ok!(Refund::_execute_refund(redeem_id, vec![0u8; 100], vec![0u8; 100]));
+
+            assert!(
```

Comment:
You're only testing 1 out of the 4 post conditions?

Reply:
Also leaving out a ton of the preconditions, but probably that is ok - the post conditions are most important

Reply:
As commented above, most of the other conditions are directly related to the functionality of other components so I'm not sure it makes sense to add here.

Reply:
Well, the ticket does say spec based _integration_ tests. I agree that in this unit test it wouldn't really make sense to check, but I would suggest making it an integration test instead